src/Runner.Worker/ActionManager.cs

```diff
@@ -1102,9 +1190,12 @@ public sealed class ScriptActionExecutionData : ActionExecutionData
     public sealed class CompositeActionExecutionData : ActionExecutionData
     {
         public override ActionExecutionType ExecutionType => ActionExecutionType.Composite;
-        public override bool HasPre => false;
-        public override bool HasPost => false;
+        public override bool HasPre => PreSteps.Count > 0;
+        public override bool HasPost => PostSteps.Count > 0;
+        public List<JobExtensionRunner> ContainerSetupSteps { get; set; }
+        public List<Pipelines.ActionStep> PreSteps { get; set; }
         public List<Pipelines.ActionStep> Steps { get; set; }
+        public Stack<Pipelines.ActionStep> PostSteps { get; set; }
```

Comment:
note to self: still trying to wrap my head around something w/ PreSteps/PostSteps... will come back to this..

Reply:
We spoke offline, following existing pattern for caching containers